### PR TITLE
I118 integrate viva

### LINKED ISSUE #118: Integrate "VIVA" Input Validator into PC2
**Feature Description**: 
Many contest judging staffs are now using "input validators" for verifying the test data for contest problems.  PC2 currently supports the ability to "define an input validator" -- that is, the user can write their own input validator and install it in PC2.  However, unlike the situation with OUTPUT validators (where in addition to supporting user-defined output validators PC2 has both the legacy PC2 validator and the CLICS validator available), there is current no "built-in" INPUT validator capability in PC2.

David van Brackle (aka "vanb"), who is the Chief Judge for both the ICPC Southeast Regional and for the North American Championship, has written a portable Input Validator and has offered it, in both Java source and compiled form, to us.  The system is called VIVA (Vanb's Input Verification Assistant).  It can be invoked from a command line, or using a GUI, or its classes can be directly invoked from within any Java 8+ program; its classes are contained in a single self-contained .jar file.

We should integrate VIVA into PC2 as an available tool for users.

**Have you considered other ways to accomplish the same thing?**   
* we could write our own Input Validator
* we could continue to not support any Input Validators other than ones which users write.

Neither of these is as attractive as integrating an existing, well-tested, maintained, Input Validator.

**Do you have any specific suggestions for how your feature would be ***implemented*** in PC^2?** 

The simplest solution would probably be to invoke VIVA with an external shell (command line), or to invoke its GUI.  The drawback of these approaches would be that the entire Input Validation process would be external to PC2.  Specifically, this probably means there is little hope of accessing the _results_ which VIVA produces in order to display those results in the existing "Input Validation Results" grid.  (More specifically, the results _might_ be able to be accessed by implementing some sort of parsing of the VIVA output, but this would likely be tedious, error-prone, and subject to future regression failures.)

A better solution would probably be to figure out how to use the VIVA classes within PC2, so that (1) VIVA _patterns_ could be defined by the user and (2) VIVA results could be displayed within PC2.  According to the VIVA documentation (available in the VIVA download from http://viva.vanb.org/), VIVA was designed to support direct class invocation from an external Java program.

**Additional context**:
VIVA is maintained as a SourceForge project; vanb has said that he's willing to give us access to the project if it would be useful to us.

VIVA is currently published as Version 2.1, although vanb has indicated that V2.2 is coming soon.  According to him there are no API differences between 2.1 and 2.2.

### PR DESCRIPTION
**Pull request by clevengr**
### Description of what the PR does
   Integrates the VIVA Input Validator into PC2, also fixing some issues with the existing Custom Input Validator code so that both VIVA and Custom Input Validators operate in a uniform manner.

### Issue which the PR fixes
Fixes #118.  The PR includes 15 new JUnit tests which test both the "VivaAdapter" class (the class used to adapt VIVA to PC2) as well as testing new ContestSnakeYAMLLoader support for loading Input Validator configurations (both VIVA and Custom Input Validator configurations) via YAML.  The PR also includes an updated version of the PC2 Admin Guide with Appendix O (Input Validators) updated to describe both how to use VIVA, how to use Custom Input Validators, and how to specify Input Validator configurations via YAML files.

### Environment in which the PR was developed (OS,IDE, Java version, etc.)
  Windows 8.1 and 10 with Java version "1.8.0_201"

### Precise steps for _testing_ the PR (i.e., how to demonstrate that it works correctly)
See the attached file [PR.i118.TestSteps.txt](https://github.com/pc2ccs/pc2v9/files/5432486/PR.i118.TestSteps.txt) for complete testing steps for both interactive Input Validator configuration as well as testing YAML loading of Input Validator configurations.

### CODE REVIEW

**Comment by troy2914** [comment]
Ugh, so many file changes,  presumably NL endings...

**Comment by clevengr** [comment]
Yes, I'd like to avoid those NL changes... any suggestions as to how?

john

**Comment by clevengr** [comment]
@troy2914 Would the "NL endings" problem(s) you referred to (above) be fixed by the outstanding PR #111 ?

**Comment by troy2914** [comment]
I am not sure, though that PR is out of date now.

**Comment by clevengr** [comment]
Why is PR #111 "out of date now"?

**Comment by troy2914** [comment]
> Why is PR #111 "out of date now"?

it's showing conflicts of a few files.  And it would be re-done after this was committed.

**Comment by clevengr** [comment]
By "after *this *was committed"... does "*this*" refer to PR #124?

**Comment by lane55** [comment]
Per Troy's comment about NL, there are tens of thousands of lines to review if the NL are not fixed.
Other than fixing the NL directly, is there a way for github to ignore NL changes.  I doubt it.

**Comment by clevengr** [comment]
> Per Troy's comment about NL, there are tens of thousands of lines to review if the NL are not fixed.
> Other than fixing the NL directly, is there a way for github to ignore NL changes. I doubt it.

My impression was that the way to do this (force github to "ignore NL changes") is to tell GIT to always store files with Unix newlines, and then to configure it so that when you check out a file it converts the NLs to the local platform (and correspondingly, when you check a file back IN it converts the NLs to Unix form).

It's also my impression that this is what PR #111 was intended to do?   This is one of the things I'd like to go over in a meeting...

**Comment by troy2914** [comment]
Looking better.   Please continue with the rest of the NL changing files.

**Comment by troy2914** [comment]
src diffs now look sane,  please fix the test ones.

**Comment by clevengr** [comment]
I believe I have finished fixing all the issues with NLs in the PR files.  Note the following:

- Two files under the "test" folder (**ui/EditProblemPaneTest** and **core/model/ProblemTest**) still show "100% of file lines changed -- _even though I converted them using dos2unix and they show as "unix" files on my machine and I have pushed them_.  I cannot explain this, unless it's some kind of delay on GitHub...
- There are still other files in my local machine Git repo that show as "Windows" files.  However, I believe that none of them affect reviewing the PR.
- The PR shows "153 files modified".  However, I should point out that _over 100 of those files_ are new "test data" files which were added under /testdata to support the 15 new JUnit tests I added for testing VIVA input validator operations.  You can spend your time reviewing all those test data files if you want, but I'm not sure why....

Please report if you see any other issues, or if you have any idea why **ProblemTest** and **EditProblemPaneTest** are still showing as "100% replacements" even though they are now in Unix format...

**Comment by troy2914** [comment]
ProblemTest.java and EditProblemPaneTest.java were in DOS (with CRLF line terminators) previously, not unix.

**Comment by clevengr** [comment]
Hmmm.... so I presume this is an issue we'll have to live with, short-term,
and which will go away once we get PR #111 completely implemented?

**Comment by troy2914** [comment]
for right now you should change those 2 files to DOS file format,  and all these NL stuff will disappear with #111

**Comment by troy2914** [comment]
The diff looks sane now.   Seems primary area to test, is to ensure the UI is properly detecting changes, and that 
input and output validators still work.

**Comment by clevengr** [comment]
The "test list" which I attached to the PR includes steps for testing
various UI changes; I think I covered all cases but of course I welcome
people trying other tests to see if they can make it misbehave.  I also
tried all combinations of VIVA input validation (with both valid and
invalid patterns) plus two different versions of a Custom Input Validator
(both of which are checked in, in the "testdata" folder).  I did NOT try
any special OUTPUT validation tests (other than all the normal ones which
run with JUnits); I'm not aware that I made any changes that would affect
Ouput Validation (but of course it's good to check/test).

And BTW, the answer to a question which came up, which several of us
*thought* we knew the answer but no one was certain:  "yes, when you push a
change to a fork which has a PR, GitHub automatically updates the files in
the PR".  I didn't have to do anything, after running dos2unix on the
offending files, other than "push" to my fork; GitHub took care of updating
the PR.  Very nice.

File: src/edu/csus/ecs/pc2/ui/DefineCustomInputValidatorPane.java
Comment (lane55) [inline src/edu/csus/ecs/pc2/ui/DefineCustomInputValidatorPane.java:281]: Should log message and displayed messages be identical?

File: test/edu/csus/ecs/pc2/imports/ccs/ContestSnakeYAMLLoaderTest.java
Comment (lane55) [inline test/edu/csus/ecs/pc2/imports/ccs/ContestSnakeYAMLLoaderTest.java:1289]: Is this a SOMEDAY or a TODO ?

**Review by lane55 — APPROVED** [review]

**Review by lane55 — CHANGES_REQUESTED** [review]
1 - In Problem class, adding fields (String [] and SF) to (flyweight) Problem may bloat/change class size.
Note that all SF and other large data structures are currently stored in ProblemDataFiles.

2 - per comment, is the SOMEDAY (aka likely never) be changed to TODO ?

File: src/edu/csus/ecs/pc2/core/model/Problem.java
Comment (clevengr) [reply src/edu/csus/ecs/pc2/core/model/Problem.java:210]: Most of the VIVA patterns I have seen are very small -- on the order of 5-10 lines of code.  I think it's reasonable to keep the pattern as a String in the Problem class.

File: src/edu/csus/ecs/pc2/core/model/Problem.java
Comment (clevengr) [reply src/edu/csus/ecs/pc2/core/model/Problem.java:464]: First comment:  I don't think Apache Commons is currently part of our vendor library; making this change would require adding that library, which I think it rather large and seems like unwarranted bloat for such a small gain.
Second comment: I'll assume that since you said "Someday", you don't consider this an item that's necessary to fix now to get the PR approved.
On that basis, I'm marking the comment "Resolved".

File: testdata/InputValidatorTest/SumitInputValidator.java
Comment (clevengr) [reply testdata/InputValidatorTest/SumitInputValidator.java:1]: I suppose it's possible this could happen, although the class is pretty simple so it seems unlikely.  I don't think this is worth holding up the approval of the PR; if you think it's necessary I think it should be filed as a separate Issue.

File: src/edu/csus/ecs/pc2/ui/RunInputValidatorsPane.java
Comment (clevengr) [reply src/edu/csus/ecs/pc2/ui/RunInputValidatorsPane.java:260]: To each their own style; your approach would require executing an extra runtime statement to assign the value of probs[prob] to the variable "problem" so that the variable could then be used 'more clearly' (in your opinion).  I'm not convinced that's a worthwhile tradeoff -- but again, to each their own style.

I'm assuming you don't think this rises to the level of rejecting the PR because of it.  If that's not the case, feel free "unresolve" the conversation and/or add a new Issue for it.  (But I don't see anyone getting around to spending their valuable time on fixing such a trivial item...)

File: src/edu/csus/ecs/pc2/ui/DefineCustomInputValidatorPane.java
Comment (clevengr) [reply src/edu/csus/ecs/pc2/ui/DefineCustomInputValidatorPane.java:281]: I don't see why this should necessarily be the case; displayed messages are generally for the user sitting at the console, while log messages are, generally speaking, more for the systems person who's going to pore through logs looking for detailed information which in fact might not make any sense to the average user at the console.

I will concede however that the systems person looking at the log might also like to see the "Internal error" part of the message, along with the "cannot enable Add/Update button"; I'll push that as a change to the PR.

File: test/edu/csus/ecs/pc2/imports/ccs/ContestSnakeYAMLLoaderTest.java
Comment (clevengr) [reply test/edu/csus/ecs/pc2/imports/ccs/ContestSnakeYAMLLoaderTest.java:1289]: First, aren't ALL "TODOs" effectively "someday"?   If it's something "to be done", it's by definition NOT currently done, so it "needs to be done", and that could be later today, tomorrow, next week, next month, next year... at some point in the future -- i.e. "someday".
Second, we do not have, have never had, any formal specification of what anything like "TODO" or "SOMEDAY" means in a comment.  Not saying we wouldn't benefit by having such; just saying we do not have any such thing and never have had any such thing.  So asking "Is this a SOMEDAY or a TODO" is pretty much a moot question since there's nothing anywhere that defines what either of those formally means.
Third, **I did not put this comment in the code**; it was there when I started working on this PR. 

I don't think this is an important issue, at least not with regard to this PR.  If you disagree, feel free to file a new Issue on it.

**Comment by clevengr** [comment]
> 1 - In Problem class, adding fields (String [] and SF) to (flyweight) Problem may bloat/change class size.
> Note that all SF and other large data structures are currently stored in ProblemDataFiles.
> 
> 2 - per comment, is the SOMEDAY (aka likely never) be changed to TODO ?

I will make the first requested change(s) and push an update.
Regarding the second requested change, see my comments on that particular comment.

Finally, I think there is an error in process here.  First, I got a message "laned Approved these changes".  If I hadn't been bogged down with other things and had acted on that message immediately, I would have taken "Approval" as meaning it was OK to merge the change (since I'm pretty sure that's what the intent of "Approving" is supposed to be).  As it is, I didn't get to processing the "Approval" before I got a subsequent message "Changes Requested".  I believe that "requesting changes" is incompatible with "Approval".  If you want changes, request them; if you're willing to accept the PR as is, Approve it -- but don't do both.

At least, that's how I think the process is supposed to work...

File: src/edu/csus/ecs/pc2/core/model/Problem.java
Comment (clevengr) [reply src/edu/csus/ecs/pc2/core/model/Problem.java:210]: If you disagree, I suggest filing this as a new Issue.

**Comment by lane55** [comment]
re: Approval to Changes requested.    I think that worked properly, I approved, then after a walk I re-viewed
the PR and the Problem and saw the flyweight class was potentially no longer flyweight.   

We should consider a refactor for the existing String [] (testCaseDataFilenames, testCaseAnswerFilenames) fields 
moving the contents out of the Problem class, in part because it is not scalable and potentially bloats the Problem.

**Comment by lane55** [comment]
Should it hold up the PR, no, I'll grant this exception.  Furture code
  should use foreach for reasons below.   

  Yes foreach this is a sytle question, and yes in my professional opinion it
  is more clear, consistent with most other code in the system, in thie case
  the overhead is negligible esp in this case, and less error prone.

**Comment by lane55** [comment]
Not a show stopper, just a note.

**Comment by lane55** [comment]
Yes all TODOs, XXX, etc. are sometday - my point is that there is 
  no agreement about how to use these notes and when they should
  be done or how urgent they are.

  The question is _when_ should someday/todo occur?   

  At work either all TODOs, etc. are fixed and removed or there
  is some way to determine what the priority of this todo is.

  JUnits should not be broken or commented out ad nauseam.  They should
  be written, removed or scheduled to be written (aka put issue number
  into the comment showing when they are fixed, for example).

**Comment by lane55** [comment]
What do you disagree with ?

  I cannot find what this refers to, so at this time I need
  more information before filing an issue or requesting
  the change in the PR.

**Comment by lane55** [comment]
Agreed.  If the pattern is changed to a String, I approve.

  Next time, I recommend encapsulating the Viva fields in a
  VivaProblemSettings class because a Problem flyweight is a flyweight,
  So only having a single class reference is better than 15+ more fields,
  also using the class makes it very clear that the fields only refer
  to Viva rather than having some affect on a Problem.  

  In the future I will not approve a PR if fields are added that
  can be naturally put into a class/encapsulated.  This isn't COBOL :)

**Comment by clevengr** [comment]
I have pushed an update to PR #124 which moves all VIVA variables and results out of the `Problem `class and into a new class `VivaInputValidatorSettings`.  The `Problem` class now contains just a single variable related to VIVA -- a reference to a `ViviInputValidatorSettings` object; all accessors which previously referenced VIVA data stored in `Problem` now delegate to methods in the `VivaInputValidatorSettings` object.  

The updated code passes all JUnits (including the ones which were added as part of the PR explicitly to test VIVA, plus all the pre-existing JUnits).   I reran the complete set of Interactive and Batch Tests described in the file  `PR.i118.TestSteps.txt` attached to the original PR #124 Description; all of them pass as expected.

Requesting re-review of the PR and (hopefully) approval.

**Review by lane55 — APPROVED** [review]
Approved.

File: src/edu/csus/ecs/pc2/core/model/Problem.java
Comment (lane55) [reply src/edu/csus/ecs/pc2/core/model/Problem.java:204]: Postpone this change to be included in a future refactor.

File: src/edu/csus/ecs/pc2/core/model/Problem.java
Comment (lane55) [reply src/edu/csus/ecs/pc2/core/model/Problem.java]: Instead of ProblemDataFile, put VIVA settings in VivaInputValidatorSettings.   VIVA settings will be stored
in Problem, until a refactor moves their storage outside of Problem similar to how ProblemDataFiles are stored.

Comment (lane55) [reply src/edu/csus/ecs/pc2/core/model/Problem.java]: With move of all Viva settings into new VivaInputValidatorSettings class, this item is resolved

**Pull request merged**
